src/embed/http.rs

```diff
@@ -1,96 +1,234 @@
 use std::convert::TryFrom;
-use std::error::Error;
-use std::fmt;
-use std::str;
 use wasm_bindgen::JsCast;
 use wasm_bindgen::JsValue;
 use wasm_bindgen_futures::JsFuture;
 use web_sys::{RequestInit, RequestMode};
 
+// s makes map_err calls nicer by mapping a error to its debug-printed string.
+fn s<D: std::fmt::Debug>(err: D) -> String {
+    format!("{:?}", err)
+}
+
+// rust_fetch fetches using a rust (as opposed to browser Fetch) http client.
+// It consumes the input request (it needs to own the request body). The response
+// returned will have status and body filled (doesn't do headers at this point).
+// Non-200 status code does not constitute an Err Result.
+//
+// TODO TLS
+// TODO timeout/abort
+// TODO log req/resp
+// TODO understand what if any tokio runtime assumptions are implied here
+pub async fn rust_fetch(
+    hyper_client: &hyper::Client<hyper::client::HttpConnector>,
+    http_req: http::Request<String>,
+) -> Result<http::Response<String>, FetchError> {
+    let (parts, req_body) = http_req.into_parts();
+    let mut builder = hyper::Request::builder()
+        .method(parts.method.as_str())
+        .uri(&parts.uri.to_string());
+    for (k, v) in parts.headers.iter() {
+        builder = builder.header(
+            k,
+            v.to_str()
+                .map_err(|e| FetchError::InvalidRequestHeader(s(e)))?,
+        );
+    }
+    let hyper_req = builder
+        .body(hyper::Body::from(req_body))
+        .map_err(|e| FetchError::InvalidRequestBody(s(e)))?;
+
+    let mut hyper_resp = hyper_client
+        .request(hyper_req)
+        .await
+        .map_err(|e| FetchError::RequestFailed(s(e)))?;
+    let http_resp_builder = http::response::Builder::new();
+    let http_resp_bytes = hyper::body::to_bytes(hyper_resp.body_mut())
+        .await
+        .map_err(|e| FetchError::ErrorReadingResponseBody(s(e)))?;
+    let http_resp_string =
```

Comment:
[nit] I guess if we're always decoding JSON we do want this as a string, but I'd probably still type the `Response::body` as `Vec<u8>` and just let callers who want it as a string do that.

Reply:
100% of our callers presently want a String so going to leave it as is